Migrate to modern bundler & React 18

echocrow opened the pull request:
Closes #652

This PR migrates the admin off of Gatsby and onto a modern bundler. Currently this is Vite, but can be swapped for webpack (TBD which fits better long-term).

Furthermore this also upgrades React to v18, along with various helper packages.

Some backwards-compatibility was put in place in the form of (bare-bone) Gatsby & Reach Router shims. I also kept the same static (`static`) & dist (`public`) folder names in place (instead of `public` and `dist`), also for backwards compatibility.

~~PR is still WIP.~~
**Update**: (hopefully) ready for review!

- [x] Migrate off Gatsby
- [x] Decide on Vite/webpack/?
- [x] Upgrade from React 16 to React 18
- [x] Migrate to Reach Router to React Router (successor of Reach Router in React 17+)
- [x] Swap old Gatsby imports
- [x] ~~Swap old Reach Router imports~~
  can complete this after this PR; shim for `@reach/router` is in place to port functionality to `react-router-dom` while allowing the old import & syntax
- [x] Upgrade or migrate from @radix-ui (no support for React 18+)
- [x] Upgrade or migrate other minor helper packages that lack React 18+ support (react-country-flag, react-datepicker, react-json-view, ...)
- [x] Fix notification panel (somewhere got bugged along the way, won't open on click)
- [x] Fix select broken routes (e.g. individual product view)

One semi-noteworthy downgrade is that now all pages are client-site rendered. On Gatsby, the index/invite/login/pw-reset pages were pre-rendered. Maybe there is a way to replicate this in Vite/webpack, but unsure if worth the effort?

On the plus site, `yarn build` went from 1.9m to 17s on my machine.
`yarn dev` goes into ready state in <300ms, but Vite cheats here (only builds upon first page request)

srindom commented:
@echocrow - this is great! Would it make sense to already move this to [medusajs/medusa](https://github.com/medusajs/medusa) instead?

echocrow commented:
> @echocrow - this is great! Would it make sense to already move this to [medusajs/medusa](https://github.com/medusajs/medusa) instead?

@srindom i'd be a bit cautious with the move to the monorepo. i think it's the right goal, but it might come with some extra requirements:

right now, people can fork this repo to add their own modifications. while this process makes it tedious to manage one's own codebase and merge upsteam update, it at least offers _some_ way.
once this gets relocated into the monorepo and packaged, I believe that functionality goes away until a plugin/extension system in is place?

unless im missing something, i'd suggest the following roadmap:
- finish migration off of gatsby (+ overall package updates/upgrades)
- implement plugin/extension system
- package and publish to npm
  - this then provides a new, more manageable way to extend the admin and make it your own
  - this also means it no longer matters where this project lives
- port to monorepo and continue development there
- freeze/archive this repo; people can still fork it, but won't receive future updates
  - (maybe archive later, and back-port some feature updates for X months, to give ppl time to transition?)

echocrow commented:
I _believe_ I worked out most kinks resulting from the various migrations, including storybook, so I think this PR is ready for review.

there is a fair but to unpack, but a summary of changes include:
- Gatsby -> Vite
  - basic shims are in place for links and navigate fn
  - site-effect: top-level invite/login/etc. pages are no longer separate pre-rendered pages; it's all one single index.html entrypoint. however, I did make sure the bundles are code-split across these routes
  - with this, there is no production "start" or "serve" command; final deployment is just a static html page (404s need to redirect to it)
- React v16 -> v18
  - reach router -> react-router dom
    - basic shims are in place for "sub-routers" (routes) & route elements
  - radix-ui v0 -> v1
  - various other packages also upgraded
    - `react-json-view` seems to be no longer maintained, warning about unmet peer deps for react <= v17; that said, seems to work fine OOTB; can leave as-is, or migrate to another JSON viewer package
- `GATSBY_MEDUSA_BACKEND_URL` -> `MEDUSA_BACKEND_URL`
  - still backwards compatible with `GATSBY_MEDUSA_BACKEND_URL` and `GATSBY_STORE_URL`
- "hi index" page replaced with spinner
- various code cleanup, in particular anything vite or react strict mode flagged, e.g.
  - removed duplicate object keys
  - made  side-effecty reducers pure
  - refactored "bad" markup (button inside button, div in tbody)

kasperkristensen reviewed (approved):
LGTM! Great work @echocrow, this is a fantastic improvement to the DX 👍

echocrow commented:
@kasperkristensen as this was already merged into `next`, does it make sense to close this PR, reopen from `next` to `develop`, and continue dev/feedback there?

kasperkristensen commented:
@echocrow Yes, we should definitely do that. As the plan is now, we will push the `next` branch to our staging environment next week, so we can start looking into getting it released as soon as possible 💪

echocrow commented:
@kasperkristensen & team: how's the vite setup + pkg upgrades been been so far?

if they're here to stay, I'll follow up on the last post-migration cleanup item I had, updating all the `@reach` import shims to the newer React router equivalents, and open as PR for `next`. 🙂

olivermrbl commented:
@echocrow That would be great! We aim to deploy the React + Vite version to our staging environment EoW or early next, so it would be nice to get those last changes in before then. Let me know, if you are keen to join in for some testing - I'd be happy to throw you an invite to our staging env.

Also, feel free to open the PR for `develop` instead of `next` 💪

echocrow commented:
exciting news! this week's a bit packed, but I should get to it over the weekend latest

> Let me know, if you are keen to join in for some testing - I'd be happy to throw you an invite to our staging env.

depends on what's needed when - do share more details here or on discord!

> Also, feel free to open the PR for develop instead of next 💪

`develop` doesn't seem to have the Vite & React 18 migrations yet, and would thus merge all of those + current next changes into this branch - but I take it that's the intend? 👀

olivermrbl commented:
> but I take it that's the intend

Exactly! 💪 

Will keep you posted re testing. When do you expect to be able to open the PR? Currently planning releases in the coming week.

echocrow commented:
> Do you have an overview of the things missing before we'd be ready test in a prod. env? Would be great with a checklist similar to the one provided in the description to this PR.

IIRC, nothing actually that's not already covered in the OP! at least nothing that should get in the way of testing.

even the cleanup task mentioned (re router elements) is not a blocker and almost more of a chore than a refactor.

background:
previously, routes would use `@reach/route` and look sth like
```jsx
import { Router, Link } from "@reach/router"

<Router>
  <MyComp1 path="foo" />
  <MyComp2 path="bar" />
</Router>
```
under react router 6, the new syntax is
```jsx
import { Routes, Route } from "react-router-dom";

<Routes>
  <Route path="foo" element={<MyComp1 />} />
  <Route path="bar" element={<MyComp2 />} />
</Routes>
```
in the current repo though, they both do effectively the exact same thing, thank to the [router shim](https://github.com/echocrow/medusa-admin/blob/vite/src/compat/reach-router-compat.tsx) I added for compatibility, injected via a [vite resolve alias](https://github.com/echocrow/medusa-admin/blob/vite/vite.config.ts) for `@reach/router`.
the remaining cleanup is just going through all instances and actually convert them to the new import and syntax - more for consistency than actual functionality.

to still provide a list though:

Short-term cleanup:
- [ ] Replace legacy `@reach/router` syntax to newer `react-router-dom` syntax.
  (pure refactor)

Long-term cleanup:
- [ ] Remove `gatsby` and `@reach/router` compatibility shims.
  (leaving them for a bit shouldn't hurt; they are light and are just there to ease migration for those who may have extended the repo but haven't switch to the new packages yet)

adding both to OP!
